helix-view/src/document.rs

```diff
@@ -209,6 +310,189 @@ impl Document {
+                let mut buf = String::with_capacity(capacity);
+                let mut total_read = 0;
+
+                // Loop until the entire chunk has been decoded into `buf`.
+                loop {
+                    let (result, read, ..) = decoder.decode_to_string(
+                        &chunk.as_bytes()[total_read..],
+                        &mut buf,
+                        is_last,
+                    );
+                    // Track how many bytes we have read so far, in case we need to allocate more
+                    // capacity to `buf`.
+                    total_read += read;
+
+                    // Check if we need to allocate more capacity to `buf`, otherwise append
+                    // to `changes`.
+                    match result {
+                        encoding_rs::CoderResult::InputEmpty => {
+                            debug_assert_eq!(total_read, chunk.len());
+                            let diff = similar::TextDiff::from_graphemes(chunk, &buf);
+                            let diff_ops = diff.grouped_ops(3);
+                            let diff_changes = diff_ops.iter().flatten().filter_map(|op| {
+                                let (tag, chunk_range, buf_range) = op.as_tag_tuple();
+                                match tag {
+                                    similar::DiffTag::Insert | similar::DiffTag::Replace => {
+                                        let new_str = buf.tokenize_graphemes()[buf_range].concat();
+                                        let start = chunk_range.start + total_text_read;
+                                        let end = chunk_range.end + total_text_read;
+                                        Some((start, end, Some(new_str.into())))
+                                    }
+                                    similar::DiffTag::Delete => {
+                                        let start = chunk_range.start + total_text_read;
+                                        let end = chunk_range.end + total_text_read;
+                                        Some((start, end, None))
+                                    }
+                                    similar::DiffTag::Equal => None,
+                                }
+                            });
+                            changes.extend(diff_changes);
+                            break;
+                        }
+                        encoding_rs::CoderResult::OutputFull => {
+                            debug_assert!(buf.len() > total_read);
+                            let needed_capacity = Self::calculate_decode_capacity(
+                                &mut decoder,
+                                &chunk.as_bytes()[total_read..],
+                            );
+                            buf.reserve(needed_capacity);
+                        }
+                    }
+                }
+                total_text_read += chunk.chars().count();
```

Comment:
`chunk.len_chars()`, it's a method on the rope

Reply:
The iterator returns `&str`, not ropes.

Reply:
Yeah, `Rope`'s chunk iterator returns `&str`.  That's the whole point, to give you direct access to the underlying chunks of utf8 data.

Having said that, it's not at all clear to me why you need to be counting the `char`s here.  You could just do everything in terms of bytes.  (Assuming I'm understanding the surrounding context well enough.)